_posts/2022-08-16-zio-http.md

```diff
@@ -0,0 +1,778 @@
+    program
+}
+```
+
+### Request => Response
+
+In `zhttp`, `Request`s are processed into `Response`s via implementations of a
+`sealed trait Http[-R, +E, -A, +B]`, which itself
+`extends (A => ZIO[R, Option[E], B])`. From the latter, we can quickly infer
+that `R` and `E` are the _resource_ and _Error_ channels of a `ZIO` effect, and
+we're going to be converting an `A` to a `B` effectually.
+
+There are some included type alias to shorten this signature, however in this
+article we will try to stick to the full version. Also, not, that due to the
+first type alias, the official documentation tends to often refer to thing as an
+"http app" or "app".
+
+```scala
+type HttpApp[-R, +E] = Http[R, E, Request, Response]
+type UHttpApp        = HttpApp[Any, Nothing]
+type RHttpApp[-R]    = HttpApp[R, Throwable]
+type UHttp[-A, +B]   = Http[Any, Nothing, A, B]
+```
+
+As a quick note, this section will have `R` as `Any` and `E` as `Nothing`. We
+will discuss including resources, and error handling later in the article.
+
+Let's take a moment to dig into our first endpoint:
+
+```scala
+  val app: Http[Any, Nothing, Request, Response] = Http.collect[Request] {
+    case Method.GET -> !! / "owls" => Response.text("Hoot!")
+  }
+```
+
+We will use `A` and `B` here, knowing that above `A = Request` and
+`B = Response`. `Http.collect[A]` is a `PartialCollect[A]` - which behaves like
+a PartialFunction, meaning we're going to pattern match on something relating to
+`A` and return a `B`.
+
+We're matching against a `Request`, so let's look closer at the `case` statement
+above. The tricky syntax is the infixed `->` operator, so let's first look to
+the immediate _right_ of it: `!! / "owls"`. This is a `Path`, and `!!` denotes
+the root of the path (an empty path, which will be important later). On the
+_left_ of `->` is `Method.GET` - a `Method`. What `->` does, is tuple2 together
+the things on the left/right of it. The definition is
+
+```scala
+@inline def -> [B](y: B): (A, B) = (self, y)
+```
+
+In our case
+
+```scala
+case Method.GET -> !! / "owls" => Response.text("Hoot!")
+```
+
+and
+
+```scala
+case (Method.GET, !! / "owls") => Response.text("Hoot!")
+```
+
+should behave identically. So what's going on, is we are looking at a `Request`
+value, and matching on it's `Method` and `Path` - if they match, we will return
+our `Response`.
+
+It will be important for later, but we can reference the request `req` in our
+response, for example, via something like
+
+```scala
+case req @ Method.GET -> !! / "owls" => Response.text("Hoot!")
+```
+
+As a quick aside, `Http.collect` also internally lifts these to `Option`s, so it
+can handle the case of `None` when nothing may match.
+
+A `Method` models an HTTP request method, i.e. `GET`, `POST`, `DELETE`, etc...
+
+A `Path` models an HTTP request path. As mentioned above, `!!` represents the a
+path root. `/` is a path delimiter that starts extraction of the left-hand side.
```

Comment:
I was working on clearing this up a bit more, but it seems in the recent release they've updated the implementation, and it doesn't function as expected/documented. 

Ultimately, I wanted to explain that `/` is a left associative operator, and `/:` is a right associative operator - useful for partial matching ( and thus a `!!` on the right could terminate that partial matching).

Using their example below, `/` and `/:` seem to end with the same results (no partial matching)

https://dream11.github.io/zio-http/docs/v1.x/dsl/request/#path

I'm going to reach out on discord and see if I can figure out what the issue is.

Reply:
It looks like it now just needs to start with an empty string for the partial matching to work, so that's working again. I feel like the behavior changed a bit, and the comments about `!!` are no longer relevant, so I've removed the confusing comments around it.